ext/random/engine_user.c

```diff
@@ -50,6 +53,7 @@ static uint64_t generate(php_random_status *status)
 			result += ((uint64_t) (unsigned char) Z_STRVAL(retval)[i]) << (8 * i);
 		}
 	} else {
+		zend_throw_exception(spl_ce_DomainException, "A random engine must return a non-empty string", 0);
```

Comment:
I would rather use a ValueError here

Reply:
I considered that, but `ValueError` is documented as:

> A ValueError is thrown when the type of an argument is correct but the value of it is incorrect.

Thus it is documented to apply to **parameters / arguments**, which is not exactly the case here.

Reply:
We could update the `ValueError` documentation; but probably throwing an `Error` directly is the way to go. `zend_throw_error()` accepts `NULL` for the first parameter, and that is already used quite a lot across php-src. I don't think that introducing a new error type would be particularly useful, since `Error`s are not supposed to be caught.